### PR TITLE
Add support for first/last for double/float/long #10702

### LINKED ISSUE #10702: DoubleFirstAggregatorFactory is not supported during ingestion for rollup
version:0.20
image

### PR DESCRIPTION
**Pull request by ankit0811**
Fixes #10702

### Description

This PR revives https://github.com/apache/druid/pull/10949 which address issue #10702 by adding support for doubleLast, doubleFirst, FloatLast, FloatFirst, longLast and longFirst.

We now support Numeric Last/First data types while ingestion and native query (for MSQ as well)
#### Ingestion
- using spec we do support aggregation behavior for Double/Float and Long last/first column types.
- this is still pending change on the UI to be supported ([reverted PR](https://github.com/apache/druid/pull/10794/files)) which will be taken part as a follow-up PR

#### Query
- Querying this data type is not yet supported in SQL as we do require some calcite changes (tobe done in a follow-up PR)
- We can use native json spec query to obtain teh numeric Last and first value from the ingested column
usage:
```
{
      "type": "doubleLast",
      "name": "a3",
      "fieldName": "last_double_added", // while ingesting this column is of type DoubleLast
      "timeColumn": "__time"
}
```

#### Release note

<hr>

##### Key changed/added classes in this PR
* `AbstractSerializableLongObjectPairSerde` abstract class to share serde between double/float/long
* `AbstractSerializablePairLongObjectBufferStore.java`
* `AbstractSerializablePairLongObjectColumnHeader.java`
* `AbstractSerializablePairLongObjectColumnSerializer.java`
* `AbstractSerializablePairLongObjectDeltaEncodedStagedSerde.java`
* `AbstractSerializablePairLongObjectSimpleStagedSerde.java`
* `SerializablePairLongDoubleComplexMetricSerde` for double
* `SerializablePairLongFloatComplexMetricSerde` for float
* `SerializablePairLongStringComplexMetricSerde` for long
* `GenericFirstAggregateCombiner` first  agg combiner to share between double/float/long
* `GenericLastAggregateCombiner` first  agg combiner to share between double/float/long
* `DoubleFirstAggregatorFactory`, `DoubleLastAggregatorFactory`
* `FloatFirstAggregatorFactory`, `FloatLastAggregatorFactory`
* `LongFirstAggregatorFactory`, `LongLastAggregatorFactory`

Further introduces new `ColumnType` 
 * serializablePairLongDouble
 * serializablePairLongFloat
 * serializablePairLongString

Post suggestions, we are no longer using `GenericIndexed`. Using the pattern defined in `SerializablePairLongStringComplexMetricSerde` we observed an improvement of 40% in terms of storage (These number are against ingestion of 1 days worth of wikiticker data)

To-Do
These will be taken up in a follow-up PR 
- Add ingestion support from UI for first/last types
- SQL compatibility for querying (latest/earliest) the new ingestion type (Today this is queryable via native json queries not via sql. Needs more discussion on making this SQL compatible)

<hr>

This PR has:

- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by abhishekagarwal87** [comment]
Thank you @ankit0811 for reviving this PR. I will review this soon.

**Comment by LakshSingla** [comment]
Thanks @ankit0811 for the PR!
This will also allow MSQ to use these aggregators for ingestion and querying. We should update the MSQ's limitations once this PR is merged in a follow-up PR (after testing).

**Review by LakshSingla — COMMENTED** [review]
Thanks for raising this PR!! I am relatively unfamiliar with the complex column's serde and am still going through the implementation, however, I have a few high-level comments, based on my understanding of the StringFirst/Last aggregators:

1. In the `deserializeColumn`, we should allocate the first byte for a version number. This would help if we decide to update the strategy down the line and want backward compatibility.

2. The current StringFirst/Last comparators use delta encoding and compression on the numeric column to reduce the size. Also, the PR (https://github.com/apache/druid/pull/12879) which introduced this change also added a few classes which help in this. I suppose that we would want it sometime. The versioning as mentioned above would help in making these changes if we decide that we don't want it now. However, the complexity of changing from one version to another is high and we require backward compatibility therefore we should question if that should be done in the original PR itself (check this comment: https://github.com/apache/druid/blob/d63eff3b1b6b15fac1bb278a784ce6c391a52ce2/processing/src/main/java/org/apache/druid/query/aggregation/SerializablePairLongStringComplexMetricSerde.java#L54-L66)

3. Can we refactor the preexisting LongString serde classes to be subclasses of the newly introduced `AbstractSerializableLongObjectPairSerde`? I see that there are some similarities in the methods. In any case, it would be helpful if all of them are under the same umbrella, and we can further categorize the newly added classes as children of `AbstractSerializableLongNumbericPairSerde`

File: processing/src/main/java/org/apache/druid/query/aggregation/SerializablePairLongDoubleComplexMetricSerde.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+      }
+
+      @Override
+      public byte[] toBytes(@Nullable SerializablePairLongDouble inPair)
+      {
+        if (inPair == null) {
+          return new byte[]{};
+        }
+
+        ByteBuffer bbuf = ByteBuffer.allocate(Long.BYTES + Byte.BYTES + Double.BYTES);
```
Comment (somu-imply) [inline processing/src/main/java/org/apache/druid/query/aggregation/SerializablePairLongDoubleComplexMetricSerde.java:87]: If it is a null do we need to allocate the byte buffer to an additional double as we are not adding anything there. While doing the get depending on the null byte we can access the double.

File: processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstAggregator.java [hunk 2]
```diff
@@ -19,30 +19,36 @@
...
+  public DoubleFirstAggregator(BaseLongColumnValueSelector timeSelector, ColumnValueSelector valueSelector, boolean needsFoldCheck)
   {
-    super(timeSelector, valueSelector);
+    super(timeSelector, valueSelector, needsFoldCheck);
     firstValue = 0;
   }
 
   @Override
-  void setCurrentValue()
+  void setFirstValue(ColumnValueSelector valueSelector)
```
Comment (somu-imply) [inline processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstAggregator.java:37]: Do we need the valueSelector here as an arg ? Since it is already available

File: processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstAggregator.java [hunk 3]
```diff
@@ -19,30 +19,36 @@
...
 import org.apache.druid.segment.BaseLongColumnValueSelector;
+import org.apache.druid.segment.ColumnValueSelector;
 
-public class DoubleFirstAggregator extends NumericFirstAggregator<BaseDoubleColumnValueSelector>
+public class DoubleFirstAggregator extends NumericFirstAggregator
 {
   double firstValue;
 
-  public DoubleFirstAggregator(BaseLongColumnValueSelector timeSelector, BaseDoubleColumnValueSelector valueSelector)
+  public DoubleFirstAggregator(BaseLongColumnValueSelector timeSelector, ColumnValueSelector valueSelector, boolean needsFoldCheck)
```
Comment (somu-imply) [inline processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstAggregator.java:30]: Does double first  need a fold check ?

File: processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java [hunk 4]
```diff
@@ -100,27 +106,37 @@ public DoubleFirstAggregatorFactory(
...
-    final BaseDoubleColumnValueSelector valueSelector = metricFactory.makeColumnValueSelector(fieldName);
+    final ColumnValueSelector valueSelector = metricFactory.makeColumnValueSelector(fieldName);
     if (valueSelector instanceof NilColumnValueSelector) {
       return NIL_BUFFER_AGGREGATOR;
     } else {
       return new DoubleFirstBufferAggregator(
           metricFactory.makeColumnValueSelector(timeColumn),
-          valueSelector
+          valueSelector,
+          StringFirstLastUtils.selectorNeedsFoldCheck(
```
Comment (somu-imply) [inline processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java:135]: This this is dealing with numeric capabilities probably we should move away from StringFirstLastUtils and create a separate utils for numeric. Also the fold check can be avoided

**Review by somu-imply — COMMENTED** [review]
Thanks for this PR. I am still going over this and will add my comments as I go over the rest.

File: processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstAggregator.java (hunk 3 above)
Comment (ankit0811) [reply processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstAggregator.java:30]: I believe so. Let me know if I am mistaken

File: processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java (hunk 4 above)
Comment (ankit0811) [reply processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java:135]: Ack. Will create a new numeric util

**Comment by ankit0811** [comment]
> Thanks for raising this PR!! I am relatively unfamiliar with the complex column's serde and am still going through the implementation, however, I have a few high-level comments, based on my understanding of the StringFirst/Last aggregators:
> 
> 1. In the `deserializeColumn`, we should allocate the first byte for a version number. This would help if we decide to update the strategy down the line and want backward compatibility.
> 2. The current StringFirst/Last comparators use delta encoding and compression on the numeric column to reduce the size. Also, the PR ([Improve String Last/First Storage Efficiency #12879](https://github.com/apache/druid/pull/12879)) which introduced this change also added a few classes which help in this. I suppose that we would want it sometime. The versioning as mentioned above would help in making these changes if we decide that we don't want it now. However, the complexity of changing from one version to another is high and we require backward compatibility therefore we should question if that should be done in the original PR itself (check this comment: https://github.com/apache/druid/blob/d63eff3b1b6b15fac1bb278a784ce6c391a52ce2/processing/src/main/java/org/apache/druid/query/aggregation/SerializablePairLongStringComplexMetricSerde.java#L54-L66
>    )
> 3. Can we refactor the preexisting LongString serde classes to be subclasses of the newly introduced `AbstractSerializableLongObjectPairSerde`? I see that there are some similarities in the methods. In any case, it would be helpful if all of them are under the same umbrella, and we can further categorize the newly added classes as children of `AbstractSerializableLongNumbericPairSerde`

Best if we take this in a separate PR? else it will be too big a PR to be reviewed. It already touches quite a few classes

**Comment by LakshSingla** [comment]
> Best if we take this in a separate PR? else it will be too big a PR to be reviewed. It already touches quite a few classes

Seems reasonable to me, we can mark this as an improvement once we get the functionality in.

File: processing/src/main/java/org/apache/druid/query/aggregation/first/FloatFirstAggregatorFactory.java [hunk 5]
```diff
@@ -281,7 +236,7 @@ public byte[] getCacheKey()
   public ColumnType getIntermediateType()
   {
     // if we don't pretend to be a primitive, group by v1 gets sad and doesn't work because no complex type serde
```
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/query/aggregation/first/FloatFirstAggregatorFactory.java:238]: This can be removed from all the numeric factories

Comment (ankit0811) [reply processing/src/main/java/org/apache/druid/query/aggregation/first/FloatFirstAggregatorFactory.java:238]: Wont this throw an exception as `AggregateFactory::getIntermediateType` default implementation calls `getType` which throws `UnsupportedOperationException`  and looks like the doc says `getType` will be deprecated in support of `getIntermediateType`

Comment (LakshSingla) [reply processing/src/main/java/org/apache/druid/query/aggregation/first/FloatFirstAggregatorFactory.java:238]: This was aimed at the comment, apologies for being unclear the first time. Currently intermediate types just return the primitive, which I got to know was a hack for making it work sometimes (after discussion with Gian). 
Now we have the correct complex types therefor we can clean out this comment.

```suggestion

```

File: processing/src/main/java/org/apache/druid/query/aggregation/last/NumericLastBufferAggregator.java [hunk 6]
```diff
@@ -22,7 +22,7 @@
 import org.apache.druid.collections.SerializablePair;
 import org.apache.druid.common.config.NullHandling;
 import org.apache.druid.query.aggregation.BufferAggregator;
-import org.apache.druid.query.monomorphicprocessing.RuntimeShapeInspector;
+import org.apache.druid.query.monomor:Q:q!phicprocessing.RuntimeShapeInspector;
```
Comment (somu-imply) [inline processing/src/main/java/org/apache/druid/query/aggregation/last/NumericLastBufferAggregator.java:25]: Seems like came from vi editor. This needs a change

**Comment by LakshSingla** [comment]
Thanks for resolving the merge conflicts and the comments. I'll test it from MSQ's perspective and check if it works fine with MSQ queries in a couple of days.

File: processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstAggregator.java (hunk 3 above)
Comment (LakshSingla) [reply processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstAggregator.java:30]: @somu-imply  I am not completely aware of the foldCheck, which implementation of the selectors produces results that cannot be optimized? Also, is there any particular reason that double won't require the check, since it seems analogous to what StringFirst et al. are doing?

**Review by LakshSingla — COMMENTED** [review]
I pulled in the patch and tested the changes out for MSQ, and the queries are now working, which didn't due to lack of serde, therefore confirming that this indeed unblocks EARLIEST and LATEST on MSQ 🚀 

Leaving a final set of review comments from my side. Thanks for being patient and receptive during the review process :)

**Comment by imply-cheddar** [comment]
This PR doesn't actually solve hte problem it needs to solve as the columns that itproduces will be extremely large compared to the expected size.  The complex column serde needs to follow the pattern followed by `SerializedPairLongStringComplexMetricSerde`.  It's not okay to use the normal `GenericIndexed` for these columns.  While I'm not against merging this just to get something in place, I am against this code existing in any release as it will mean that some people might actually create segments with incredibly large columns, not realizing why, and hten we have to carry forward the bad versions of the code to read those columns back out all because we just didn't implement it according to the already existing pattern the first time around.

**Comment by imply-cheddar** [comment]
Generally speaking, I don't believe that we should merge the code with the `GenericIndexed` method of generating the columns.  That's going to create the need to always support that.  We have a better way to store things (the way that the String versions do it) and we shoudl follow those instead.  If we really want to do these in two PRs, I'd suggest leaving this PR as is and create another PR on top of it to switch the serde around and then merge them in quick succession.

**Comment by clintropolis** [comment]
>There is a better way that the stringLast/First aggregators already use, can we please adjust the ComplexMetricSerde implementations to align with the pattern followed in SerializablePairLongStringComplexMetricSerde instead? Specifically in how that implements getSerializer() and deserializeColumn(). There are code paths for compatibility that use the GenericIndexed stuff, the pattern that we want to follow is the ones that use SerializableLongPairStringColumnSerializer and SerializablePairLongStringComplexColumn.Builder respectively.

Random drive-by comment, but I don't really think that is the best pattern to use here either though, since the pairs here are fixed width and so a much more appropriate way of compressing them that doesn't involve storing offsets separately could be used...

**Comment by ankit0811** [comment]
@imply-cheddar a qq for clarifications since this part of code is new to me

1. the current `SerializablePairLongStringComplexColumn.Builder` has a version check. I am assuming this is not required for long numeric type implementation. is the assumption correct here?

**Comment by LakshSingla** [comment]
@ankit0811 
```
1   {version=v1, timestamp=2013-08-31T00:00:00.000Z, event={continent=Asia, earliest_user=masterYi, latest_user=stringer, double_first_delta=111.0, double_last_delta=-9.0, long_first_delta=111, long_last_delta=-9, float_first_delta=111.0, float_    last_delta=-9.0}}],
  1 {version=v1, timestamp=2013-08-31T00:00:00.000Z, event={continent=Asia, double_first_delta=null, long_last_delta=null, double_last_delta=null, float_first_delta=null, earliest_user=masterYi, long_first_delta=null, latest_user=stringer, float    _last_delta=null}}]
```

There are some failing ITs. Can you please take a look at them? Looks like you should be able to identify the culprit after looking at the output diff. (A very wild guess - the new metric isn't hooked up properly perhaps? )

If you need help setting up the ITs for debugging, then please reach out, and I should be able to help with it.

We do want to get this change in, and a green build would expedite the process.

**Comment by ankit0811** [comment]
@LakshSingla this seems to be cos of some changes wrt vectorization. Will try and address this in a couple of days.

**Comment by daniel-imgarena** [comment]
Will we see this in Druid 28? I see myself back on the need for a doubleLast or longFirst for way too long

**Comment by LakshSingla** [comment]
@daniel-imgarena The branch for Druid 28 has been cut, and only release blockers & regressions are allowed at this point. Unfortunately, we won't be seeing this in the 28 release, since it's a new feature.

File: processing/src/main/java/org/apache/druid/query/aggregation/SerializablePairLongStringColumnSerializer.java
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/query/aggregation/SerializablePairLongStringColumnSerializer.java:1]: This should also inherit from the `AbstractSerializablePairLongObjectColumnSerializer` class.

File: processing/src/main/java/org/apache/druid/query/aggregation/AbstractSerializablePairLongObjectBufferStore.java
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/query/aggregation/AbstractSerializablePairLongObjectBufferStore.java:1]: The String methods could also inherit from this class I suppose.

**Review by LakshSingla — COMMENTED** [review]
Leaving a partial review as of now, but there are two main comments, apart from the line items:
1. A lot of the abstract methods for serde & aggregators can be merged with the string's methods without any tweaking (I suppose), since they don't rely on the numeric properties of these aggregators. We should do that, otherwise, there will be a lot of code duplication. Also, one would need to keep track of both the classes (abstract ones & the string ones) while fixing bugs or making any improvements. 
2. Thanks for taking the time to rewrite without using the GenericIndexed. I am curious if there's any benchmarking that you did or if there was a performance/size benefit that you observed from the change.

File: processing/src/main/java/org/apache/druid/query/aggregation/SerializablePairLongDoubleComplexMetricSerde.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+import org.apache.druid.segment.column.ColumnBuilder;
+import org.apache.druid.segment.data.ObjectStrategy;
+import org.apache.druid.segment.serde.cell.NativeClearedByteBufferProvider;
+import org.apache.druid.segment.writeout.SegmentWriteOutMedium;
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+import java.util.Comparator;
+
+public class SerializablePairLongDoubleComplexMetricSerde extends AbstractSerializableLongObjectPairSerde<SerializablePairLongDouble>
```
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/query/aggregation/SerializablePairLongDoubleComplexMetricSerde.java:33]: There should be a version associated with the serde, at the first position Check out the string implementation which does this. This will help in future iterations of the serde, without making any backward incompatible changes.

File: processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java [hunk 8]
```diff
@@ -142,12 +160,24 @@ public VectorAggregator factorizeVector(
...
+          valueSelector,
+          capabilities.toColumnType(),
+          ColumnType.DOUBLE
+      );
+      return new DoubleFirstVectorAggregator(timeSelector, objectSelector);
+    }
+
+    VectorObjectSelector vSelector = columnSelectorFactory.makeObjectSelector(fieldName);
+    if (capabilities != null) {
+      return new DoubleFirstVectorAggregator(timeSelector, vSelector);
```
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java:180]: Why is this check required? I don't think we are using the capabilities inside the aggregator

File: processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java [hunk 9]
```diff
@@ -142,12 +160,24 @@ public VectorAggregator factorizeVector(
...
-      return new DoubleFirstVectorAggregator(timeSelector, valueSelector);
+      VectorObjectSelector objectSelector = ExpressionVectorSelectors.castValueSelectorToObject(
+          columnSelectorFactory.getReadableVectorInspector(),
+          fieldName,
+          valueSelector,
+          capabilities.toColumnType(),
+          ColumnType.DOUBLE
+      );
+      return new DoubleFirstVectorAggregator(timeSelector, objectSelector);
+    }
```
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java:176]: I am not sure if this is correct. If the type is numeric, why do we need an explicit cast to `Double`. I could be wrong, but this seems kinda suspicious.

File: processing/src/main/java/org/apache/druid/query/aggregation/first/NumericFirstVectorAggregator.java [hunk 10]
```diff
@@ -164,7 +215,7 @@ void updateTimeWithNull(ByteBuffer buf, int position, long time)
    * Abstract function which needs to be overridden by subclasses to set the
    * latest value in the buffer depending on the datatype
    */
-  abstract void putValue(ByteBuffer buf, int position, int index);
+  abstract void putValue(ByteBuffer buf, int position, Number number);
```
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/query/aggregation/first/NumericFirstVectorAggregator.java:218]: Let's revert this change if we can. It will lead to some duplication, however, this will lead to the boxing of the primitives, which is what the column selectors aim to reduce (if they can). Therefore there are separate methods like `isNull` + `getLong` which are preferred instead of `getObject` because the former doesn't auto-box the primitive variable to an object.

File: processing/src/main/java/org/apache/druid/query/aggregation/first/NumericFirstVectorAggregator.java [hunk 11]
```diff
@@ -132,14 +183,14 @@ public void aggregate(
    *
    * @param buf         byte buffer storing the byte array representation of the aggregate
    * @param position    offset within the byte buffer at which the current aggregate value is stored
-   * @param time        the time to be updated in the buffer as the last time
-   * @param index       the index of the vectorized vector which is the last value
+   * @param time        the time to be updated in the buffer as the first time
+   * @param number      number which is the first value
    */
-  void updateTimeWithValue(ByteBuffer buf, int position, long time, int index)
+  void updateTimeWithValue(ByteBuffer buf, int position, long time, Number number)
```
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/query/aggregation/first/NumericFirstVectorAggregator.java:189]: Autoboxing alert, as mentioned in the other comment. Let's see if we can get away without this change.

Comment (LakshSingla) [reply processing/src/main/java/org/apache/druid/query/aggregation/first/NumericFirstVectorAggregator.java:189]: It might not be as clean as the current code, however, there shouldn't be a lot of duplication.

File: processing/src/main/java/org/apache/druid/query/aggregation/first/StringFirstAggregatorFactory.java
[the remark this answers is not in the document]
Comment (ankit0811) [reply processing/src/main/java/org/apache/druid/query/aggregation/first/StringFirstAggregatorFactory.java:1]: Sure. I can modify the PR to make StringFirst and Last use the Generic/Abstract classes.

File: processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java (hunk 9 above)
Comment (LakshSingla) [reply processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java:176]: With the string factory, we'd need to check if it is of a numeric type and cast if so. But if it is a numeric, then it shouldn't be the case, at least for DOUBLE objects. Also, if it is a String aggregator, now we return a DoubleFirstVectorAggregator (if capabilities correspond to String capabilities).

File: processing/src/main/java/org/apache/druid/query/aggregation/AbstractSerializablePairLongObjectSimpleStagedSerde.java [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    Object rhsObject = value.getRhs();
+
+    return new StorableBuffer()
+    {
+      @Override
+      public void store(ByteBuffer byteBuffer)
+      {
+        Preconditions.checkNotNull(value.getLhs(), String.format(Locale.ENGLISH, "Long in %s must be non-null", pairCLass.getSimpleName()));
```
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/query/aggregation/AbstractSerializablePairLongObjectSimpleStagedSerde.java:58]: Don't need to use String.format. 
```suggestion
        Preconditions.checkNotNull(value.getLhs(), "Long in %s must be non-null", pairCLass.getSimpleName());
```

File: processing/src/main/java/org/apache/druid/query/aggregation/AbstractSerializablePairLongObjectSimpleStagedSerde.java [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+        if (rhsObject != null) {
+          if (pairCLass.isAssignableFrom(SerializablePairLongLong.class)) {
+            rhsBytes = Long.BYTES;
+          } else if (pairCLass.isAssignableFrom(SerializablePairLongDouble.class)) {
+            rhsBytes = Double.BYTES;
+          } else if (pairCLass.isAssignableFrom(SerializablePairLongFloat.class)) {
+            rhsBytes = Float.BYTES;
+          }
+        }
+        return Long.BYTES + Byte.BYTES + rhsBytes;
```
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/query/aggregation/AbstractSerializablePairLongObjectSimpleStagedSerde.java:88]: This computation is incorrect. For case when rhsObject is null, we'd only be storing `Long.BYTES + Byte.BYTES`, so this should be conditional.

Comment (LakshSingla) [reply processing/src/main/java/org/apache/druid/query/aggregation/AbstractSerializablePairLongObjectSimpleStagedSerde.java:88]:
```suggestion
        return Long.BYTES + Byte.BYTES + (rhsObject == null ? 0 : rhsBytes);
```

Comment (ankit0811) [reply processing/src/main/java/org/apache/druid/query/aggregation/AbstractSerializablePairLongObjectSimpleStagedSerde.java:88]: rhsBytes is assigned to 0 and is assigned value only is `rhsBytes` is non null. So isn't it already conditional?

**Comment by ankit0811** [comment]
@LakshSingla thank you so much for reviewing this PR and being patient throughout the process. Have tried to address your comments let me know if theres any thing more left

**Review by LakshSingla — APPROVED** [review]

**Comment by LakshSingla** [comment]
Thanks for the patch @ankit0811.

I think there's some follow up that can be done, in this area
1. Make sure string pairs work as expected with null & empty strings - This has been a regression before the patch, and I have raised a PR for the same
2. Make the EARLIEST and the LATEST work with SQL
3. Add tests for MSQ
4. Look for some cleaner abstractions

**Comment by LakshSingla** [comment]
Can you add a release note for the patch

File: processing/src/test/java/org/apache/druid/query/aggregation/first/DoubleFirstVectorAggregationTest.java [hunk 14]
```diff
@@ -86,39 +91,80 @@ public long[] getLongVector()
       @Override
       public boolean[] getNullVector()
       {
-        return NULLS;
+        return null;
       }
     };
-    selector = new BaseDoubleVectorValueSelector(new NoFilterVectorOffset(VALUES.length, 0, VALUES.length)
+    selector = new VectorObjectSelector()
```
[the remark this answers is not in the document]
Comment (ankit0811) [reply processing/src/test/java/org/apache/druid/query/aggregation/first/DoubleFirstVectorAggregationTest.java:97]: Yes. It will be required as the selector can now be of either VectorObject or VectorValue.

[the remark this answers is not in the document]
Comment (LakshSingla) [reply processing/src/test/java/org/apache/druid/query/aggregation/first/DoubleFirstVectorAggregationTest.java:97]: Thanks for the explanation

**Comment by LakshSingla** [comment]
I still believe that we can clean up the code a bit further if we get rid of the SerializablePairLongString abstractions and keep them as generics, however, I don't know what entails such a change. That would allow the reuse of multiple classes. However, that will be taken in a follow up PR, so its a go-ahead from my side.

**Comment by cryptoe** [comment]
SInce we have identified the caveats, we can merge this and work on it in a follow up PR. 
Thanks @ankit0811 and @LakshSingla for the PR and reviews.

**Review by cryptoe — APPROVED** [review]
